pkg/pattern/tokenization/replacer.go

```diff
@@ -0,0 +1,859 @@
+
+import (
+	"bytes"
+	"slices"
+)
+
+var (
+	placeholderNumber    = []byte("<NUM>")
+	placeholderHex       = []byte("<HEX>")
+	placeholderUUID      = []byte("<UUID>")
+	placeholderTimestamp = []byte("<TIMESTAMP>")
+	placeholderDuration  = []byte("<DURATION>")
+	placeholderBytesize  = []byte("<BYTESIZE>")
+	placeholderIP        = []byte("<IP>")
+)
+
+// Integer numbers after these words won't be replaced by `<NUM>`.
+var numPreservingKeys = [][]byte{[]byte("status"), []byte("status_code"), []byte("httpStatus")}
+
+const (
+	maxYear   = 2100 // anything above that is unlikely to be a timestamp...
+	minHexLen = 12   // 48 bits min for free-standing hex strings (e.g "0123456789ab") or 42 bits for "0xABCDEF1234" strings
+)
+
+var boundaryChars = [256]bool{}
+
+func init() {
+	for i := 0; i < 128; i++ { // for now, we don't consider non-ASCII chars as boundary
+		if i < '0' || (i > '9' && i < 'A') || (i > 'Z' && i < 'a') || i > 'z' {
+			boundaryChars[i] = true
+		}
+	}
+
+	// We need these keys sorted in an ascending length to efficiently compare them.
+	slices.SortStableFunc(numPreservingKeys, func(a, b []byte) int {
+		return len(a) - len(b)
+	})
+}
+
+type replacer struct {
+	source, dest []byte
+
+	// This is the last position that was copied to the destination buffer.
+	tail int
+	// This is the current position we are working with
+	cur int
+	// This is essentially used for lookahed operations
+	head int
+
+	// Here's some ASCII art to visualize that (the long string of dashes
+	// visualizes the log line:
+	//
+	// 0 <already in dest, copied or replaced >   `tail`   <about to be copied or replaced>   `cur`   <lookahead>   `head`   <remaining unprocessed log>   `len(source)`
+	// |---------------------------------------------|------------------------------------------|---------------------|------------------------------------------|
+
+	// A somewhat hacky solution that allows us to preserve specific numeric
+	// values we care about, like status and status_code.
+	preserveNextNumbers bool
+}
+
+// commit advances the current position marker to the lookahead position, i.e.
+// we are committing to consume everything we've looked ahead so far.
+func (r *replacer) commit() {
+	r.cur = r.head
+}
+
+func (r *replacer) resetHead() {
+	r.head = r.cur
+}
+
+func (r *replacer) resetHeadExpr() bool {
+	r.head = r.cur
+	return true // useful when included in boolean expressions with advance()
+}
+
+func (r *replacer) backtrack() {
+	r.head--
+}
+
+func (r *replacer) consumeUpToCurrent() {
+	r.tail = r.cur
+}
+
+// advanceUintRet returns the actual value of the number we read, as well as its
+// length. The value might overflow for large numbers, so it's also important to
+// check the length!
+func (r *replacer) advanceUintRet() (val uint, length uint) {
+	var c byte
+	for ; r.head < len(r.source); r.head++ {
+		c = r.source[r.head]
+		if c < '0' || '9' < c {
+			break
+		}
+		length++
+		val = val*10 + uint(c-'0') // TODO: use bitwise trick?
```

Comment:
This is what I meant by "bitwise trick": https://www.guidodiepen.nl/2017/03/clever-bit-tricks-with-ascii/#integervaluesinascii

I doubt it's actually faster on modern CPU architectures (and it's certainly less readable), but worth benchmarking to confirm, just in case :sweat_smile: